pkg/controller/iam/managed.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotServiceAccount)
+	}
+
+	getRequest := e.serviceAccounts.Get(e.rrn.ResourceName(cr))
+	fromProvider, err := getRequest.Context(ctx).Do()
+	if err != nil {
+		if gcp.IsErrorNotFound(err) {
+			return managed.ExternalObservation{ResourceExists: false}, nil
+		}
+		return managed.ExternalObservation{}, errors.Wrap(err, errGet)
+	}
+
+	populateCRFromProvider(cr, fromProvider)
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  isUpToDate(&cr.Spec.ForProvider, fromProvider),
+		ConnectionDetails: managed.ConnectionDetails{},
+	}, nil
+}
+
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/create
+// Note that the metadata.Name from the Kubernetes custom resource is used as the AccountID parameter
+// All other API methods use the external-name annotation
+// (set via the RelativeResourceNameAsExternalName Initializer)
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotServiceAccount)
+	}
+
+	csar := &iamv1.CreateServiceAccountRequest{
+		AccountId: cr.GetName(),
+		ServiceAccount: &iamv1.ServiceAccount{
+			DisplayName: gcp.StringValue(cr.Spec.ForProvider.DisplayName),
+			Description: gcp.StringValue(cr.Spec.ForProvider.Description),
+		},
+	}
+
+	// The first parameter to the Create method is the resource name of the GCP project
+	// where the service account should be created
+	req := e.serviceAccounts.Create(e.rrn.ProjectName(), csar)
+	fromProvider, err := req.Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
+	}
+	populateCRFromProvider(cr, fromProvider)
+	return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
+}
+
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/patch
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotServiceAccount)
+	}
+
+	sa := &iamv1.ServiceAccount{}
+	populateProviderFromCR(sa, cr)
+	psar := &iamv1.PatchServiceAccountRequest{
+		ServiceAccount: sa,
+		UpdateMask:     "description,displayName",
+	}
+	req := e.serviceAccounts.Patch(e.rrn.ResourceName(cr), psar)
+	// we don't pay attention to the result of the patch request because it is only guaranteed to contain
+	// `description` and `displayName` ie the fields we are trying to change
+	_, err := req.Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
+	}
+	return managed.ExternalUpdate{}, nil
```

Comment:
Nit: This could be abbreviated:

```go
_, err := e.serviceAccounts.Patch(e.rrn.ResourceName(cr), psar).Context(ctx).Do()
return managed.ExternalUpdate{}, errors.Wrap(err, errUpdate)
```

`errors.Wrap()` will return `nil` if `err == nil`.

Reply:
I like separating the request builder from executing it. I'm having a hard time justifying why, I think it's that I want the part that is interacting with the network to be really clearly delineated, whereas chaining it onto the end of the request builder lumps it all together.